From 34a3ef05e0180ee5b0ece2c8accd94a22b230f12 Mon Sep 17 00:00:00 2001
From: LeoLiu-oc <LeoLiu-oc@zhaoxin.com>
Date: Fri, 23 Jul 2021 15:02:04 +0800
Subject: [PATCH] ALSA: hda: Add Zhaoxin SB HDAC non snoop
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 34a3ef05e0180ee5b0ece2c8accd94a22b230f12


zhaoxin inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I40QDN
CVE: NA

----------------------------------------------------------------

Add Zhaoxin SB HDAC non snoop path support.

Signed-off-by: LeoLiu-oc <LeoLiu-oc@zhaoxin.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 sound/pci/hda/hda_intel.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/sound/pci/hda/hda_intel.c b/sound/pci/hda/hda_intel.c
index 4c8b281c3992..8ff7902ec6e6 100644
--- a/sound/pci/hda/hda_intel.c
+++ b/sound/pci/hda/hda_intel.c
@@ -240,8 +240,8 @@ MODULE_SUPPORTED_DEVICE("{{Intel, ICH6},"
 			 "{ATI, RV770},"
 			 "{VIA, VT8251},"
 			 "{VIA, VT8237A},"
-			 "{SiS, SIS966},"
-			 "{ULI, M5461}}");
+			 "{ULI, M5461},"
+			 "{ZX, ZhaoxinHDA}}");
 MODULE_DESCRIPTION("Intel HDA driver");
 
 #if defined(CONFIG_PM) && defined(CONFIG_VGA_SWITCHEROO)
@@ -1744,6 +1744,15 @@ static void azx_check_snoop_available(struct azx *chip)
 			snoop = false;
 	}
 
+	if (azx_get_snoop_type(chip) == AZX_SNOOP_TYPE_NONE &&
+		chip->driver_type == AZX_DRIVER_ZHAOXIN) {
+		u8 val1;
+
+		pci_read_config_byte(chip->pci, 0x42, &val1);
+		if (!(val1 & 0x80) && chip->pci->revision == 0x20)
+			snoop = false;
+	}
+
 	if (chip->driver_caps & AZX_DCAPS_SNOOP_OFF)
 		snoop = false;
 
-- 
2.26.2

